.teamcity/Gradle_Check/model/BucketProvider.kt

```diff
@@ -0,0 +1,239 @@
+package Gradle_Check.model
+
+import com.alibaba.fastjson.JSON
+import com.alibaba.fastjson.JSONArray
+import com.alibaba.fastjson.JSONObject
+import configurations.FunctionalTest
+import model.BuildTypeBucket
+import model.CIBuildModel
+import model.GradleSubProject
+import model.Stage
+import model.TestCoverage
+import model.TestType
+import java.io.File
+
+typealias BuildProjectToSubProjectTestClassTimes = Map<String, Map<String, List<TestClassTime>>>
+
+interface GradleBuildBucketProvider {
+    fun createFunctionalTestsFor(stage: Stage, testConfig: TestCoverage): List<FunctionalTest>
+
+    fun createDeferredFunctionalTestsFor(stage: Stage): List<FunctionalTest>
+}
+
+class StatisticBasedGradleBuildBucketProvider(private val model: CIBuildModel, testTimeDataJson: File) : GradleBuildBucketProvider {
+    private val subProjectProvider = GradleSubProjectProvider(model.subProjects)
+    private val buckets: Map<TestCoverage, List<BuildTypeBucket>> = buildBuckets(testTimeDataJson, model)
+
+    override fun createFunctionalTestsFor(stage: Stage, testCoverage: TestCoverage): List<FunctionalTest> {
+        return buckets.getValue(testCoverage).map { it.createFunctionalTestsFor(model, stage, testCoverage) }
+    }
+
+    override fun createDeferredFunctionalTestsFor(stage: Stage): List<FunctionalTest> {
```

Comment:
Previously, the code which configures deferred functional tests were scattered in different places. This function simplify the logic and put them together.

Reply:
:+1: